uwuno/src/main/java/com/learning/uwuno/controllers/playerController.java

```diff
@@ -0,0 +1,106 @@
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class playerController {
+    // Runs and injects container service
+    @Autowired
+    private roomContainerService containerService;
+
+    // GETS
+    // Returns a list of players in given room uid.
+    @GetMapping(value = "rooms/{uid}/players")
+    public ArrayList<player> players(@PathVariable String uid) {
+        if (StringUtils.isNumeric(uid)) {
+            try {
+                room room = containerService.getRoom(Integer.parseInt(uid));
+                return room.getPlayers();
+            } catch (NoSuchElementException e) {
+                throw new errorNotFound();
+            }
+        }
+        throw new errorNotFound();
+    }
+
+    // POSTS
+    // Adds a new player to the given room uid.
+    @PostMapping(value = "rooms/{uid}/players")
+    public void addPlayer(@RequestBody player newPlayer, @PathVariable String uid) {
+        if (StringUtils.isNumeric(uid)) {
+            try {
+                // TODO: Check if name already exists
+                // TODO: Generate unique id
+                room room = containerService.getRoom(Integer.parseInt(uid));
+                room.addPlayer(newPlayer);
+            } catch (NoSuchElementException e) {
+                throw new errorNotFound();
+            }
+        } else {
+            throw new errorNotFound();
+        }
+    }
+
+    // PUTS
+    // Update player name in given room uid
+    @PutMapping(value = "rooms/{uid}/players")
+    public void updatePlayerName(@RequestBody String json, @PathVariable String uid)
+            throws JsonMappingException, JsonProcessingException {
+        if (StringUtils.isNumeric(uid)) {
+            ObjectMapper mapper = new ObjectMapper();
+            try {
+                // TODO: Check if name already exists
+                room room = containerService.getRoom(Integer.parseInt(uid));
+                Map<String, String> map = mapper.readValue(json, new TypeReference<Map<String, String>>(){});
+                player player = room.getPlayer(Integer.parseInt(map.get("pid")));
+                player.setName(map.get("newName"));
+            }
+            catch (NoSuchElementException e) {
+                throw new errorNotFound();
+            }
+        } else {
+            throw new errorNotFound();
+        }
+    }
+
+    // DELETES
+    @DeleteMapping(value = "rooms/{uid}/players")
+    public void deletePlayer(@RequestBody String json, @PathVariable String uid)
+            throws JsonMappingException, JsonProcessingException {
```

Comment:
Is this done through spring's exception handling?

Reply:
No, mapper needed it.

Reply:
We still need to handle that with a server error message (I think it was a 500 something?)

Reply:
Added internal error exception class.